### PR TITLE
add `flbit-field`

### PR DESCRIPTION
**Pull request by mflatt**
The `flbit-field` operation is analogous to `bitwise-bit-field`, but it extracts bits from a flonum.

This operation seems useful to the [Herbie](https://herbie.uwplse.org/) developers, particularly at a use site where the flonum argument can be unboxed and the result is a fixnum, so that it can work without allocating.

### CODE REVIEW

**Comment by mflatt** [comment]
@mnieper — Thanks for your comments.

I don't think this operation is like the functions in 2.8 of the R6RS library, which operate on bytevectors. I don't immediately see a point to new operations on bytevectors, because it's already straightforward (using integer operations) to inspect bits of a floating point number as encoded in a bytevector.

Your point about different IEEE formats is well taken in the context a standard, and I agree that `flieee-single-bit-field` and `flieee-double-bit-field` could be sensible names when "flonum" is more loosely defined. In Chez Scheme, though, a flonum is defined to be an IEEE 64-bit floating-point number, so there's no ambiguity in `flbit-field`. Also, the usefulness of this function seems tied to performance characteristics of a specific implementation, so I'm not sure it would be a great candidate for a standard. My preference would be to stick with the more obvious name `flbit-field` for Chez Scheme, but I don't feel too strongly about this if others would prefer a longer name.

**Comment by mnieper** [comment]
Thank you for your response.

> @mnieper — Thanks for your comments.
> 
> I don't think this operation is like the functions in 2.8 of the R6RS library, which operate on bytevectors. I don't immediately see a point to new operations on bytevectors, because it's already straightforward (using integer operations) to inspect bits of a floating point number as encoded in a bytevector.

I didn't mean to add new bytevector operations; as you write, there is no point in doing so as there is already a straightforward solution.

I referred to bytevectors to have a template for naming conventions and semantics (single/double, endianess, domain of argument).

> Your point about different IEEE formats is well taken in the context a standard, and I agree that `flieee-single-bit-field` and `flieee-double-bit-field` could be sensible names when "flonum" is more loosely defined. In Chez Scheme, though, a flonum is defined to be an IEEE 64-bit floating-point number, so there's no ambiguity in `flbit-field`. Also, the usefulness of this function seems tied to performance characteristics of a specific implementation, so I'm not sure it would be a great candidate for a standard. My preference would be to stick with the more obvious name `flbit-field` for Chez Scheme, but I don't feel too strongly about this if others would prefer a longer name.

The R6RS standard way to achieve what `flbit-field` does would be to create a bytevector holding a single flonum and then extracting the bits.  This is likely non-optimised on most implementations (because an extra heap object is allocated).  Thus, for almost all implementations, a specialised `flbit-field` can be implemented more efficiently.

If it is useful for external code, a portable feature would even be better.

The root problem is, of course, that one does not simply want to cast a flonum into an exact integer (to do bitwise arithmetic with that one) because 64-bits are typically more than the fixnum range (and thus would produce an allocation again) so one needs operations acting directly on flonums/inexact reals.

In any case, the name `flbit-field` probably wouldn't clash with a future standard because that would have to include "ieee-single" and "ieee-double" for consistency.

PS How is the endianness handled in your operation?

**Comment by mflatt** [comment]
> PS How is the endianness handled in your operation?

Like `bitwise-bit-field`, there's no question of endianness. In the implementation, a flonum's value is sitting in a floating-point register, that register's bits are transferred to an integer register (with the bit order intact, so that the sign bit is still the high bit, for example), and bits are then accessed from the integer register.

**Comment by mnieper** [comment]
> Like `bitwise-bit-field`, there's no question of endianness. In the implementation, a flonum's value is sitting in a floating-point register, that register's bits are transferred to an integer register (with the bit order intact, so that the sign bit is still the high bit, for example), and bits are then accessed from the integer register.

Ah, that was my fallacy.  It is up to the `flbit-field` procedure to do the conversion in case on some supported architecture the endianness of the double and integer format does not coincide, so the user won't notice.

File: csug/numeric.stex [hunk 1]
```diff
@@ -636,6 +636,31 @@ the number.
...
 
+%----------------------------------------------------------------------------
+\entryheader
+\formdef{flbit-field}{\categoryprocedure}{(flbit-field \var{flonum} \var{exint_1} \var{exint_2})}
+\returns a nonnegative exact integer
+\listlibraries
+\endentryheader
+
+\var{flonum} must be a flonum, \var{exint_1} and \var{exint_2} must be
+exact integers in the range \scheme{0} to \scheme{64} (inclsuive), and
```
[the remark this answers is not in the document]
Comment (mflatt) [reply csug/numeric.stex]: Fixed - thanks!

**Comment by LiberalArtist** [comment]
By coincidence, I was looking for something like this last week while I was contemplating the WASM generic numeric operation [`fcopysign`](https://webassembly.github.io/spec/core/exec/numerics.html#op-fcopysign), which corresponds to the concrete instructions `f32.copysign` and `f64.copysign`. It is supposed to look inside even the bit pattern of a NaN: I wrote a little more [here](https://gitlab.com/spritely/guile-hoot/-/issues/334). (I'm still not sure why this is a useful instruction to have.)

I used an intermediate bytevector (from Racket), but I noticed that Chez Scheme provides `decode-float`. I think `flbit-field` would make it possible to replace the C implementation with Scheme, if that seemed like a good idea:

https://github.com/cisco/ChezScheme/blob/95ee804d741e5af9c41f59145b4f42877fa35ae5/c/prim5.c#L319-L322

https://github.com/cisco/ChezScheme/blob/95ee804d741e5af9c41f59145b4f42877fa35ae5/c/number.c#L1284-L1321

**Pull request merged**